### PR TITLE
App Rating Library

### PR DESCRIPTION
**Pull request by manthan-25**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,72 @@
-# react-native-app-rater
\ No newline at end of file
+
+# react-native-app-rater
+
+## Getting started
+
+`$ npm install react-native-rate-app --save`
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 2]
```diff
@@ -1 +1,72 @@
...
+
+### Mostly automatic installation
+
+`$ react-native link react-native-rate-app`
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 3]
```diff
@@ -1 +1,72 @@
...
+
+### Manual installation
+
+
+#### iOS
+
+1. In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
+2. Go to `node_modules` ➜ `react-native-rate-app` and add `RNRateApp.xcodeproj`
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 4]
```diff
@@ -1 +1,72 @@
...
+3. In XCode, in the project navigator, select your project. Add `libRNRateApp.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`
+4. Run your project (`Cmd+R`)<
+
+#### Android
+
+1. Open up `android/app/src/main/java/[...]/MainActivity.java`
+  - Add `import com.shellmonger.reactnative.RNRateAppPackage;` to the imports at the top of the file
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 5]
```diff
@@ -1 +1,72 @@
...
+  - Add `new RNRateAppPackage()` to the list returned by the `getPackages()` method
+2. Append the following lines to `android/settings.gradle`:
+  	```
+  	include ':react-native-rate-app'
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 6]
```diff
@@ -1 +1,72 @@
...
+  	project(':react-native-rate-app').projectDir = new File(rootProject.projectDir, 	'../node_modules/react-native-rate-app/android')
+  	```
+3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
+  	```
+      compile project(':react-native-rate-app')
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 7]
```diff
@@ -1 +1,72 @@
...
+  	```
+
+#### Windows
```
Comment (VishwajeetDagar) [inline README.md]: remove windows setup

File: README.md [hunk 8]
```diff
@@ -1 +1,72 @@
...
+
+1. In Visual Studio add the `RNRateApp.sln` in `node_modules/react-native-rate-app/windows/RNRateApp.sln` folder to their solution, reference from their app.
+2. Open up your `MainPage.cs` app
+  - Add `using Rate.App.RNRateApp;` to the usings at the top of the file
+  - Add `new RNRateAppPackage()` to the `List<IReactPackage>` returned by the `Packages` method
+
+
+## Usage
+```javascript
+import RNRateApp from 'react-native-rate-app';
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 9]
```diff
@@ -1 +1,72 @@
...
+}
+```
+## Props and Usage
+Prop | Description | Type | Default
+------ | ------ | ------ | ------
+ **`type`** | Rating Type: either `0`, `1` or `2` - 0 - Not Displayed, 1 - Ratings with Stars, 2 - Ratings with Emojis | number | 0
+ **`noOfDays`** | Show Ratings next after 'noOfDays' Days | number | 90
+ **`onDismiss`** | A function called when Ratings is closed or not displayed (as Days Passed < noOfDays or type is 0) | function | (none)
+ **`shouldAlwaysShow`** | When true the ratings will always be shown regardless of the noOfDays passed | boolean | `false`
+ **`thanksScreenTimeout`** | timeout (in ms) after which the ThankYouView will be closed (automatically) | number | 3000 (3s)
```
Comment (VishwajeetDagar) [inline README.md]: change this

File: README.md [hunk 10]
```diff
@@ -1 +1,72 @@
...
+ **`sendEvent`** | function to handle GA-Events Call | function | (none)
```
Comment (VishwajeetDagar) [inline README.md]: add parameters

File: package.json [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [
+    "react-native"
+  ],
+  "author": "",
+  "license": "",
+  "peerDependencies": {
+    "react-native": "^0.41.2",
```
Comment (VishwajeetDagar) [inline package.json]: remove ^

File: src/RatingComponent.js [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+      thanksVisible: false,
+      showButton: false,
+      showInputText: false,
+      rating: 1,
+      feedback: ''
+    }
+  }
+
+  componentDidMount(){
+    this.props.sendEvent({ type: 'ratings' })
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: make 'rating' string type to a meaningful string like 'RATING_MOUNTED' or 'RATING OPENED'

File: src/RatingComponent.js [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+      >
+        <TouchableOpacity
+          activeOpacity={1}
+          onPressOut={() => {
+            this.onClose(true)
+          }}
+          style={{ justifyContent: 'flex-end', flex: 1, backgroundColor: setAlpha('#000000', 50) }}
+        >
+          {rateVisible && (
+            <KeyBoardAvoidView behavior="padding">
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js:223]: Segregrate the Components into small functional components

File: src/RatingComponent.js [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+  onClose = (later = false, maxRatings = false) => {
+    const { dismiss, storeLink, noOfDays } = this.props
+    let rating = 0
+    if (later === false) {
+      // eslint-disable-next-line prefer-destructuring
+      rating = this.state.rating
+    }
+    if (maxRatings === true) {
+      Linking.canOpenURL(storeLink).then(
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: Make it in a function and put it in last

File: src/RatingComponent.js [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+        (supported) => {
+          // eslint-disable-next-line no-unused-expressions
+          supported && Linking.openURL(storeLink)
+        },
+        err => console.log(err)
+      )
+    }
+    setShowDate(rating, noOfDays)
+    if (later === true) {
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js:136]: Please add comment to explain what condition does.

File: src/RatingComponent.js [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+      clearTimeout(this.timer)
+      this.setState({ rateVisible: false, thanksVisible: false }, dismiss)
+      if(!this.state.thanksVisible) {
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: Add this functionality above setState function

File: src/RatingComponent.js [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+
+  startTimer = () => {
+    const { timeout } = this.props
+    this.timer = setTimeout(this.closeThankYouScreen, timeout)
+  }
+
+  showButton = () =>
+    this.setState({
+      showButton: true,
+      showInputText: false
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: We can remove these two state variables and check on just maxRating state.

File: src/RatingComponent.js [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+                )}
+                <TouchableOpacity style={{ marginTop: 16 }} onPress={() => this.onClose(true)}>
+                  <Text style={styles.later}>Remind me Later</Text>
+                </TouchableOpacity>
+              </TouchableOpacity>
+            </KeyBoardAvoidView>
+          )}
+          {thanksVisible && (
+            <View
+              activeOpacity={1}
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: remove active opacity

File: src/RatingComponent.js [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+
+RatingComponent.defaultProps = {
+  type: 0,
+  timeout: 1000,
+  noOfDays: 90,
+  sendEvent: () => {}
+}
+
+RatingComponent.propTypes = {
+  dismiss: PropTypes.func.isRequired,
```
Comment (jitendralakhmani-proptiger) [inline src/RatingComponent.js]: We can make is an optional prop

File: src/Ratings.js [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  dismissRatingCard = () => {
+    this.setState({ showRatingComponent: false }, this.props.onDismiss)
+  }
+
+  async shouldShow() {
+    const { onDismiss, type } = this.props
+    if( type === 0 ) {
+      onDismiss()
```
Comment (jitendralakhmani-proptiger) [inline src/Ratings.js]: experiment type should be handled from the project not the library

File: src/utils.js [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+  return color.substring(0, 7) + alphaValue // Format: #rrggbbaa
+}
+
+export function getCurrentTime() {
+  return (new Date().getTime() / msPerDay)
+}
+
+export function setShowDate(ratings = 0, noOfDays) {
+  const today = getCurrentTime()
+  const nextTime = today + noOfDays
```
Comment (jitendralakhmani-proptiger) [inline src/utils.js]: We can store unix milli sec

File: src/Ratings.js [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    const showDate = await AsyncStorage.getItem('SHOW_DATE')
+    if (showDate) {
+      const { nextTime, neverShow } = JSON.parse(showDate)
+      if (neverShow === false) {
+        const currentTime = getCurrentTime()
+        if (currentTime >= nextTime) {
+          this.setState({ showRatingComponent: true })
+        } else {
+          onDismiss()
```
Comment (jitendralakhmani-proptiger) [inline src/Ratings.js]: Dismiss function should not be called again n again. Need to figure out a solution for this.

File: src/Ratings.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    } else {
+      this.setState({ showRatingComponent: true })
+    }
+  }
+
+  render() {
+    const { type, sendEvent, storeLink, noOfDays, thanksScreenTimeout } = this.props
+    const { showRatingComponent } = this.state
+    return (
+      <View>
```
Comment (jitendralakhmani-proptiger) [inline src/Ratings.js]: return null when cond is false.

File: src/RatingsDisplay.js [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    )
+  }
+
+  render() {
+    return (
+      <View style={{ width: '100%', marginTop: 16, zIndex: 0, marginBottom: 16 }}>
+        <FlatList
+          data={this.state.data}
+          scrollEnabled={false}
+          contentContainerStyle={[styles.flatlistContainer, { zIndex: 1 }]}
```
Comment (jitendralakhmani-proptiger) [inline src/RatingsDisplay.js]: please check for zIndex

File: src/RatingsDisplay.js [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+      data = emojiSrc.map(img => img.unselected)
+    }
+    return { data, opacity }
+  }
+
+  changeRatings = (index) => {
+    const { data, opacity } = this.state
+    const { showButton, showInput, setRating, type } = this.props
+    if( type === 1 ) {
+      for (let i = 0; i <= 4; i += 1) {
```
Comment (jitendralakhmani-proptiger) [inline src/RatingsDisplay.js]: Use map instead of loops

**Review by jitendralakhmani-proptiger — APPROVED** [review]

**Review by jitendralakhmani-proptiger — APPROVED** [review]

**Pull request merged**